### PR TITLE
Optimize reflection of F# types, part 2

### PR DESCRIPTION
**Pull request by kerams**
Continuation of #9714.

- `PreComputeRecordConstructor`
- `PreComputeRecordFieldReader`
- `PreComputeRecordReader` - reworked from last time
- `PreComputeUnionConstructor`
- `PreComputeUnionReader`
- `PreComputeUnionTagReader`

`PreComputeTupleReader` and `PreComputeTupleConstructor` are a little more complicated, so perhaps another time.

The rest of the `PreCompute` family only return corresponding `System.Reflection.MethodBase` for further reflection use.

|                        Type |     Method |       Mean |      Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------------- |----------- |-----------:|-----------:|----------:|-------:|------:|------:|----------:|
| PreComputeRecordConstructor | Reflection | 554.210 ns |  3.3056 ns | 3.0921 ns | 0.0134 |     - |     - |     112 B |
| PreComputeRecordConstructor |   Compiled |   9.999 ns |  0.0416 ns | 0.0389 ns | 0.0057 |     - |     - |      48 B |
|                             |            |            |            |           |        |       |       |           |
| PreComputeRecordFieldReader | Reflection | 112.941 ns |  1.9830 ns | 1.8549 ns | 0.0029 |     - |     - |      24 B |
| PreComputeRecordFieldReader |   Compiled |   3.431 ns |  0.1047 ns | 0.1120 ns | 0.0029 |     - |     - |      24 B |
|                             |            |            |            |           |        |       |       |           |
|      PreComputeRecordReader | Reflection | 562.094 ns | 10.2099 ns | 9.5503 ns | 0.0134 |     - |     - |     112 B |
|      PreComputeRecordReader |   Compiled |  22.212 ns |  0.2661 ns | 0.2489 ns | 0.0134 |     - |     - |     112 B |
|                             |            |            |            |           |        |       |       |           |
|  PreComputeUnionConstructor | Reflection | 473.935 ns |  5.1988 ns | 4.8629 ns | 0.0105 |     - |     - |      88 B |
|  PreComputeUnionConstructor |   Compiled |   7.737 ns |  0.0595 ns | 0.0497 ns | 0.0048 |     - |     - |      40 B |
|                             |            |            |            |           |        |       |       |           |
|       PreComputeUnionReader | Reflection | 318.424 ns |  4.2661 ns | 3.9905 ns | 0.0086 |     - |     - |      72 B |
|       PreComputeUnionReader |   Compiled |  13.706 ns |  0.3290 ns | 0.4822 ns | 0.0086 |     - |     - |      72 B |
|                             |            |            |            |           |        |       |       |           |
|    PreComputeUnionTagReader | Reflection | 373.670 ns |  2.4971 ns | 2.3358 ns | 0.0029 |     - |     - |      24 B |
|    PreComputeUnionTagReader |   Compiled |   6.567 ns |  0.1600 ns | 0.2136 ns |      - |     - |     - |         - |

### CODE REVIEW

File: src/fsharp/FSharp.Core/reflect.fs [hunk 1]
```diff
@@ -286,9 +373,22 @@ module internal Impl =
                 match getInstancePropertyReader (typ, "Tag", bindingFlags) with
                 | Some reader -> (fun (obj: obj) -> reader obj :?> int)
                 | None ->
-                    (fun (obj: obj) ->
-                        let m2b = typ.GetMethod("GetTag", BindingFlags.Static ||| bindingFlags, null, [| typ |], null)
-                        m2b.Invoke(null, [|obj|]) :?> int)
+                    let m2b = typ.GetMethod("GetTag", BindingFlags.Static ||| bindingFlags, null, [| typ |], null)
+                    (fun (obj: obj) -> m2b.Invoke(null, [|obj|]) :?> int)
```
Comment (kerams) [inline src/fsharp/FSharp.Core/reflect.fs:377]: There's probably no need to look up the method on every invocation. I didn't know how to test this specifically though. When is it ever the case that a DU doesn't have a `Tag` property?

Comment (cartermp) [reply src/fsharp/FSharp.Core/reflect.fs:377]: A DU always has a tag as far as I know - struct/ref type, single case/multi case all have tag properties.

Comment (kerams) [reply src/fsharp/FSharp.Core/reflect.fs:377]: Hmm, then I'm not sure if this branch executes.

Comment (KevinRansom) [reply src/fsharp/FSharp.Core/reflect.fs:377]: I think Paul had PR to do tagless DUs a few (lot of) years ago.

**Comment by kerams** [comment]
@cartermp, what loop? The only loops I see are those that are used to create expression trees during the precomputation. There are no loops inside the compiled delegates.

**Comment by cartermp** [comment]
In this case I'm referring to a loop like this: https://github.com/dotnet/fsharp/pull/9784/files#diff-54378f53a8612b2adf50a6efb115ba5aR134

The compiler rewrites this to use an enumerator no matter which kind of `for` loop you use, see:

[current impl](https://sharplab.io/#v2:DYLgZgzgNAJiDUAfA9gBwKYDsAEBlAnhAC7oC2AsAFBpZ6EmkB0ASumMOgMZECWymVGjgLEyjADI9MAR0YBRAB6oATuggQ+mCFSoci2TslKoeHAKqZNAYQCGEdFf7FlAV27JlAMReZO2ABSk6EQAFsgwAJKYYMgg2ACywWGR0cgAlNgAvFTYudh62EGh4QAKNso2pBBZhUnhUTGMAOLBZRVF6MrV/mk5eQWo5ZUAgsoV+DWKKmoa/IxtlcGdAUT4GMhgADzIAEYAVgDaALoAfFDYAESD7Re9lH25BehKylkPeblTquqaEpU7MBsm28vm2+2O512exOJwC7w+Hy+M1+jkwADdOvp/PCEbikT85rZgMA4ZRceTyUVkg1kFAcRTcQd6QyKTFXtdKlEYM9sFJsAAGbCMRi1YowBZVCRYADmoWwAFpsABGbAwZDMlksgY1KmlIaSg4c0hc55HHRkzWWxEvZGE/gY5RYjVWln42aYRijcbDTicGYBI1emz4c5ulFOIg2TD6I0mhRpOkWl3J7Coeb6pbKAAqa3QzpTuQyIFhYf4+YZZqTVoT5Y+q3WWyhJxrVYpgbGweZdw+zxUjEcxlM6ACaSAA===)
[for x = 0 to loop](https://sharplab.io/#v2:DYLgZgzgNAJiDUAfA9gBwKYDsAEBlAnhAC7oC2AsAFBpZ6EmkB0ASumMOgMZECWymVGjgLEyjADI9MAR0YBRAB6oATuggQ+mCFSoci2TslKoeHAKqZNAYQCGEdFf7FlAV27JlAMReZO2ABSk6EQAFsgwAJKYYMgg2ACywWGR0cgAlNgAvFTYudh62EGh4QAKNso2pBBZhUnhUTGMAOLBZRVF6MrV/mk5eQWo5ZUAgsoV+DWKKmoa/IxtlcGdAUT4GMhgADzIAEYAVgDaALoAfFDYAESD7Re9lH25BehKylkPeblTquqaEpU7MBsm28vm2+2O512exOJwC7w+Hy+M1+jkwADdOvp/PCEbikT85rZgMA4ZRceTyUVkg1kFAcRTcQd6QyKTFXtdKlEYM8agAGbBEZC1YowBZVCRYADmoWwAFpsABGbAwZDMlksgY1KmlIbig4c0hc55HHRk9XmxEvZGE/gY5RYtUWln42aYRijcbDTicGYBA0emz4c4ulFOIg2TD6A1GhRpOlmp2J7Coea6pbKAAqa3QjqTuQyIFhIf4uYZJoTFrjpY+q3WWyhJyrFYp/rGgeZdw+zxUjEcxlM6ACaSAA==)

Rewriting it to a while loop and manually bumping the indexer will remove that [like so](https://sharplab.io/#v2:DYLgZgzgNAJiDUAfA9gBwKYDsAEBlAnhAC7oC2AsAFBpZ6EmkB0ASumMOgMZECWymVGjgLEyjADI9MAR0YBRAB6oATuggQ+mCFSoci2TslKoeHAKqZNAYQCGEdFf7FlAV27JlAMReZO2ABSk6EQAFsgwAJKYYMgg2ACywWGR0cgAlNgAvFTYudh62EGh4QAKNso2pBBZhUnhUTGMAOLBZRVF6MrV/mk5eQWo5ZUAgsoV+DWKKmoa/IxtlcGdAUT4GMhgADzIAEYAVgDaALoAfFDYAESD7Re9lH25BehKylkPeblTquqaEpU7MBsm28vm2+2O512exOJwC7w+Hy+M1+jkwADdOvp/PCEbikT85rZgMA4ZRceTyUVkg1kFAcRTcQd6QyKQVSC4iDYdhxsNdKlEYM8agAGZks3EAdxCpnQvKGpAFQs2tWKMAWVQkWAA5qFsDBkGLxazgryalTSvKIIwDnyFZhBQojjoyUbXZ8XsjCfwMcosYa3RT8bNMIxRuNhpxODMArawzZ8OcgyinJzMPpbYqFGk6S6AwHUPN5UtlAAVNbof157AZECwpP8St5xv5+WZ7CbAC0cvabfg2AAjM3cU7c67s0PsKt1lsoSdx6OKbGxvHmXcPs8VIxHMYZQE0kA=) but it's unlikely to matter much from a perf standpoint. Just kind of weird that trivially changing the `for` loop had no effect when it usually does.

**Comment by kerams** [comment]
Yeah, I'd be all for altering the loop were this happening inside of the delegate. Don't think it matters at all during prep stage.

**Comment by abelbraaksma** [comment]
In hot paths it matters for ints, it's about 4x slower. That  may not matter much here, though. There's an issue open to better optimize this, the example from @cartermp will help there.

**Comment by KevinRansom** [comment]
This func is compileUnionCaseConstructorFunc  ... it should not happen frequently.

**Comment by Daniel-Svensson** [comment]
@kerams if you want to further improve performance by 6-7 (on netcore 3.1) times for the "get all properties" case you can modify the expression to initialize the array in reverse order to avoid array bounds check. 

I have a method added to your benchmark from the previous PR and results laying around at home that I can post next week if it is of interest. 
But it might be regarded as a micro optimization and not too relevant here since one might expect  Expression.InitArray to be as fast as possible.

**Comment by kerams** [comment]
Are you saying 

```csharp
var a = int[5];
a[0]= 0;
a[0]= 0;
a[0]= 0;
```

**Comment by Daniel-Svensson** [comment]
@kerams actually I belive I must have a bug in my fsharp (cannot check it this week), it was a quick hack late before bedtime. 
If the length is unknown to the jit the reverse order is much faster but thinking more about it i would not expect such large improvements but only double digit percentage. There are many such "hacks" in the runtime

**Pull request merged**